resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/EntityOutputStream.java

```diff
@@ -0,0 +1,304 @@
+            throw new IllegalStateException(Messages.MESSAGES.streamIsClosed());
+        }
+        getDelegate(1).write(b);
+    }
+
+    @Override
+    public void write(final byte[] b) throws IOException {
+        if (closed.get()) {
+            throw new IllegalStateException(Messages.MESSAGES.streamIsClosed());
+        }
+        getDelegate(b.length).write(b);
+    }
+
+    @Override
+    public void write(final byte[] b, final int off, final int len) throws IOException {
+        if (closed.get()) {
+            throw new IllegalStateException(Messages.MESSAGES.streamIsClosed());
+        }
+        getDelegate(len).write(b, off, len);
+    }
+
+    @Override
+    public void flush() throws IOException {
+        getDelegate().flush();
+    }
+
+    @Override
+    public void close() throws IOException {
+        getDelegate().close();
+        closed.set(true);
+    }
+
+    /**
+     * Creates an input stream from this output stream.
+     *
+     * @return an input stream for reading the output
+     *
+     * @throws IOException if an error occurs creating the input stream
+     */
+    InputStream toInputStream() throws IOException {
+        if (!closed.get()) {
+            throw Messages.MESSAGES.streamNotClosed(this);
+        }
+        if (!exported.compareAndSet(false, true)) {
+            throw Messages.MESSAGES.alreadyExported();
+        }
+        synchronized (lock) {
+            if (file != null) {
+                final InputStream in = Files.newInputStream(file);
+                return new EntityInputStream(in, CLEANER.register(in, new FileCleaner(file)));
+            }
+            try {
+                return new ByteArrayInputStream(inMemory.toByteArray());
+            } finally {
+                inMemory.reset();
+            }
+        }
+    }
+
+    /**
+     * Creates an entity for the Apache HTTP Client based on the output stream.
+     *
+     * @return an entity for the output stream
+     */
+    AbstractHttpEntity toEntity() {
+        if (!closed.get()) {
+            throw Messages.MESSAGES.streamNotClosed(this);
+        }
+        if (!exported.compareAndSet(false, true)) {
+            throw Messages.MESSAGES.alreadyExported();
+        }
+        synchronized (lock) {
+            if (file != null) {
+                final AbstractHttpEntity result = new FileEntity(file.toFile());
+                CLEANER.register(result, new FileCleaner(file));
+                return result;
+            }
+            try {
+                return new ByteArrayEntity(inMemory.toByteArray());
+            } finally {
+                inMemory.reset();
+            }
+        }
+    }
+
+    /**
+     * The length of the length of the content written.
+     *
+     * @return the length of the content written
+     *
+     * @throws IOException if there is an error determining the length of the content
+     */
+    long getContentLength() throws IOException {
+        final Path file = this.file;
```

Comment:
redundant line?

Reply:
Maybe I should lock, but the idea was to get the current value as its volatile and then use that below.

Reply:
I see, then lock seems safer if that matters, you already use plenty of that in this class.